ringct: MLSAG speedup and zero-hash check

SarangNoether opened the pull request:
- Removes redundant point conversions in MLSAG verification
- Replaces public key hash-to-point zero checks with round hash-to-scalar zero checks
- Updates tests
- Improves error messages

SarangNoether commented:
I can revert those if that's what we really want. Will be adding some additional speedups to this, so don't merge yet.

moneromooo-monero commented:
I think I'm just easily annoyed at comparing things just to work out that there's no actual difference in code, so it feels like busywork :P Don't mind me, I'll go grumble in a corner.

SarangNoether commented:
Added a zero-check to key images per discussion with @b-g-goodell, which replaces existing public key hash checks.

vtnerd commented on src/ringct/rctSigs.cpp:
```diff
@@ -163,14 +163,21 @@ namespace rct {
-    //   the signer knows a secret key for each row in that column
-    // Ver verifies that the MG sig was created correctly        
+    // Hash a key to p3 representation
+    static void hash_to_p3(const key &k, ge_p3 &hash8_p3) {
```
This now aliases a function in `src/ringct/rctOps.cpp` which itself is an alias for a function in `src/crypto/crypto.cpp`.

Merge one of these functions somewhow. The existing one in `src/ringct/rctOps.cpp` can call this function and then immediately convert to bytes (assuming that code somewhere is still using that function).

This can also return-by-value in its current usage, although the benefit seems low since `ge_p1p1_to_p3` takes an out variable anyway.

luigi1111 replied:
I think this function should be in rctOps and probably replace hashToPoint there.

However, I'm also willing to merge anyway and leave it for the future.

SarangNoether replied:
Updated. Review requested.

vtnerd commented on src/ringct/rctSigs.cpp:
```diff
@@ -260,53 +267,58 @@ namespace rct {
-                CHECK_AND_ASSERT_MES(!(Hi == rct::identity()), false, "Data hashed to point at infinity");
-                addKeys3(R, rv.ss[i][j], Hi, c_old, Ip[j].k);
+
+                // Compute R directly
```
Mainly to ensure I didn't misread this code - the check here was only used as input to the hash function. So if it were the identity point, it couldn't manipulate any of the other calculations ... ?

SarangNoether replied:
A forgery using hashed-to-zero public keys would require that the adversary "trapdoor" both the left and right hash inputs for a given index. We can remove the checks on each hashed public key and replace them with a single check on the key image; same effect, but cheaper.

vtnerd replied:
The response wasn't exactly what I was getting at - if `R` were used in some math operation the situation might be different. But instead it is only being fed into a hash function.

SarangNoether replied:
Yes, `R` is only used for its byte representation to be fed into the hash function. I meant that the checks on line 308 have been removed and replaced with a single check on the key image instead; the speedup comes from the portion starting at line 313, where we avoid the `hashToPoint` byte conversion that's undone by `addKeys3`.

SarangNoether commented on src/ringct/rctSigs.cpp:
```diff
@@ -317,6 +328,7 @@ namespace rct {
                 toHash[ndsRows + 2 * ii + 2] = L;
             }
             c = hash_to_scalar(toHash);
+            CHECK_AND_ASSERT_MES(!(c == rct::zero()), false, "Bad signature hash");
```
It could remove the signer's need to know a private key.

vtnerd replied:
But this value is directly from a hash function - the attacker would have to find a pre-image mapping to zero. The check is probably worth it I suppose, because this would break the implementation in the next iteration (nothing is being added on line 326).

SarangNoether replied:
Yes, it would require a preimage of zero, making it extremely unlikely. One of the Bulletproofs auditors pointed out a similar check was missing in Schnorr signature verification code, and it seemed wise (and cheap) to add it here as well.

SarangNoether commented on src/ringct/rctSigs.cpp:
```diff
@@ -296,17 +301,23 @@ namespace rct {
+                hash_to_p3(pk[i][j], hash8_p3);
+                ge_p2 R_p2;
+                ge_double_scalarmult_precomp_vartime(&R_p2, rv.ss[i][j].bytes, &hash8_p3, c_old.bytes, Ip[j].k);
+                key R;
```
Cleaned up in new commit.

moneromooo-monero commented:
Can you squash the cleanup patch into the relevant commit please ?

SarangNoether commented:
This has been updated to replace the existing hash-to-point functions with the new hash-to-p3 function. I also removed `hashToPointSimple` and moved its functionality directly into the unit test that verifies the construction of the `H` Pedersen generator; it shouldn't be used elsewhere anyway.

moneromooo-monero commented on src/ringct/rctSigs.cpp:
```diff
@@ -215,7 +206,8 @@ namespace rct {
             else {
-              Hi = hashToPoint(pk[index][i]);
+              hash_to_p3(Hi_p3, pk[index][i]);
+              ge_p3_tobytes(Hi.bytes, &Hi_p3);
```
Why not keep hashToPoint, whch uses hash_to_p3 internally ? Avoids these "inlining".

notmike-5 replied:
if it isn't broke why fix it?

SarangNoether replied:
Certainly could if desired. This is used infrequently.

SarangNoether commented:
Results from a test machine show verification time in `test_ringct_mlsag` decreases from 13025 us to 12467 us, a 4.3% improvement.

fuwa0529 commented on src/ringct/rctOps.cpp:
```diff
@@ -620,44 +620,16 @@ namespace rct {
-        ge_p3_tobytes(pointk.bytes, &res);
-    }    
+    // Hash a key to p3 representation
+    void hash_to_p3(ge_p3 &hash8_p3, const key &k) {
```
Why not just returning a result? This looks very nonfunctional.
